src/prototype_rust_sdk/workflow_context.rs

```diff
@@ -218,21 +186,188 @@ impl WfContext {
     }
 }
 
-struct WFCommandFut {
+/// A Future that can be cancelled.
+/// Used in the prototype SDK for cancelling operations like timers and activities.
+pub trait CancellableFuture<T>: Future<Output = T> {
+    /// Cancel this Future
+    fn cancel(&self, cx: &WfContext);
+}
+
+struct WFCommandFut<T>
+where
+    T: From<UnblockEvent>,
+{
+    /// We need a field that references the template type T, this works
+    unused: PhantomData<T>,
+    cancellable_id: CancellableID,
     result_rx: oneshot::Receiver<UnblockEvent>,
 }
 
-impl WFCommandFut {
-    fn new() -> (Self, oneshot::Sender<UnblockEvent>) {
+impl<T> WFCommandFut<T>
+where
+    T: From<UnblockEvent>,
+{
+    fn new(cancellable_id: CancellableID) -> (Self, oneshot::Sender<UnblockEvent>) {
         let (tx, rx) = oneshot::channel();
-        (Self { result_rx: rx }, tx)
+        (
+            Self {
+                cancellable_id,
+                result_rx: rx,
+                unused: PhantomData,
+            },
+            tx,
+        )
     }
 }
 
-impl Future for WFCommandFut {
-    type Output = UnblockEvent;
+impl<T> Unpin for WFCommandFut<T> where T: From<UnblockEvent> {}
+
+impl<T> Future for WFCommandFut<T>
+where
+    T: From<UnblockEvent>,
+{
+    type Output = T;
 
     fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
-        self.result_rx.poll_unpin(cx).map(|x| x.unwrap())
+        self.result_rx.poll_unpin(cx).map(|x| x.unwrap().into())
+    }
+}
+
+impl<T> CancellableFuture<T> for WFCommandFut<T>
+where
+    T: From<UnblockEvent>,
+{
+    fn cancel(&self, cx: &WfContext) {
```

Comment:
In principle, the futures could probably store a ref to the `WfContext` and then it doesn't need to be passed here, but We don't really have to bother with that right away

Reply:
let do that in the ... future 😉